drivers/SmartThings/zigbee-window-treatment/src/aqara/aqara_utils.lua

```diff
@@ -0,0 +1,212 @@
+local PRIVATE_CLUSTER_ID = 0xFCC0
+local PRIVATE_ATTRIBUTE_ID = 0x0009
+local MFG_CODE = 0x115F
+local PREF_ATTRIBUTE_ID = 0x0401
+local SHADE_STATE_ATTRIBUTE_ID = 0x0404
+
+local SHADE_STATE_STOP = 0
+local SHADE_STATE_OPEN = 1
+local SHADE_STATE_CLOSE = 2
+local INITIALIZED_STATE = "initializedState"
+
+local PREF_REVERSE_DEFAULT = "\x00\x02\x00\x00\x00\x00\x00"
+local PREF_REVERSE_OFF = "\x00\x02\x00\x00\x00\x00\x00"
+local PREF_REVERSE_ON = "\x00\x02\x00\x01\x00\x00\x00"
+
+local FINGERPRINTS = {
+  { mfr = "LUMI", model = "lumi.curtain", device_profile = "curtain" },
+  { mfr = "LUMI", model = "lumi.curtain.v1", device_profile = "curtain" },
+  { mfr = "LUMI", model = "lumi.curtain.aq2", device_profile = "roller-shade" }
+}
+
+local aqara_utils = {}
+
+local function read_private_attribute(device, cluster_id, attribute_id)
+  local message = cluster_base.read_attribute(device, data_types.ClusterId(cluster_id), attribute_id)
+  message.body.zcl_header.frame_ctrl:set_mfg_specific()
+  message.body.zcl_header.mfg_code = data_types.validate_or_build_type(MFG_CODE, data_types.Uint16, "mfg_code")
+  device:send(message)
+end
+
+local function write_private_attribute(device, cluster_id, attribute_id, dt, value)
+  device:send(cluster_base.write_manufacturer_specific_attribute(device, cluster_id, attribute_id, MFG_CODE,
+    dt, value))
+end
+
+local function enable_private_cluster_attribute(device)
+  write_private_attribute(device, PRIVATE_CLUSTER_ID, PRIVATE_ATTRIBUTE_ID, data_types.Uint8, 1)
+end
+
+local function read_pref_attribute(device)
+  read_private_attribute(device, Basic.ID, PREF_ATTRIBUTE_ID)
+end
+
+local function write_pref_attribute(device, str)
+  write_private_attribute(device, Basic.ID, PREF_ATTRIBUTE_ID, data_types.CharString, str)
+end
+
+local function write_reverse_pref_default(device, str)
+  write_pref_attribute(device, PREF_REVERSE_DEFAULT)
+end
+
+local function write_reverse_pref_off(device, str)
+  write_pref_attribute(device, PREF_REVERSE_OFF)
+end
+
+local function write_reverse_pref_on(device, str)
+  write_pref_attribute(device, PREF_REVERSE_ON)
+end
+
+local function send_lift_percentage_cmd(device, command, level)
+  device:send_to_component(command.component, WindowCovering.server.commands.GoToLiftPercentage(device, level))
+end
+
+local function send_open_cmd(device, command)
+  send_lift_percentage_cmd(device, command, 100)
+end
+
+local function send_close_cmd(device, command)
+  send_lift_percentage_cmd(device, command, 0)
+end
+
+local function emit_shade_state_event(device, shadeLevel)
+  if shadeLevel == 100 then
+    device:emit_event(capabilities.windowShade.windowShade.open())
+  elseif shadeLevel == 0 then
+    device:emit_event(capabilities.windowShade.windowShade.closed())
+  else
+    device:emit_event(capabilities.windowShade.windowShade.partially_open())
+  end
+end
+
+local function emit_shade_level_event(device, level)
+  device:emit_event(capabilities.windowShadeLevel.shadeLevel(level))
+end
```

Comment:
Why are there so many one-line functions? I don't think this significantly increases readability.

Reply:
I'll edit and push.